fix: Check for Kubernetes connection before retrieving pods

cdrage opened the pull request:
fix: Check for Kubernetes connection before retrieving pods

### What does this PR do?

* Gets rid of the ECONNREFUSED connection errors when first starting
  Podman Desktop with a Podman Machine running
* Adds a check (listing namespaces) for an active connection before
  running `kubernetesListPods`.
* Moves podsInfos.set(result) to higher up as sometimes we are not
  checking Kubernetes if it's not active

### Screenshot/screencast of this PR

<!-- Please include a screenshot or a screencast explaining what is doing this PR -->

N/A

### What issues does this PR fix or reference?

<!-- Please include any related issue from Podman Desktop repository (or from another issue tracker).
-->

PR created while investigating: https://github.com/containers/podman-desktop/issues/2096

### How to test this PR?

<!-- Please explain steps to reproduce -->

Start Podman Desktop with developer console and a podman machine running with kind STOPPED. You should no longer see
any ECONNREFUSED when first starting Podman Desktop.

Signed-off-by: Charlie Drage <charlie@charliedrage.com>

jeffmaury reviewed (commented):
Still having an error message on Windows. I do believe this kind of test/hidding should not be done at the UI level but rather in the Kubernetes client

cdrage commented:
> Still having an error message on Windows. I do believe this kind of test/hidding should not be done at the UI level but rather in the Kubernetes client

What's the error / log so I can see where else it's getting a connection refused?

cdrage commented:
> Still having an error message on Windows. I do believe this kind of test/hidding should not be done at the UI level but rather in the Kubernetes client

For how we do it in listPods, we just check to see if there is a connection, and if not, then return an empty array.

```ts
  async listPods(): Promise<PodInfo[]> {
    const pods = await Promise.all(
      Array.from(this.internalProviders.values()).map(async provider => {
        try {
          if (!provider.libpodApi) {
            return [];
          }
 ```

IMHO, I think we should have a `checkConnection()` boolean function for Kubernetes that will check to see if we're connected and return true / false.

I'll update this PR.

cdrage commented:
Ready for another review!

* Found another section where we do not check the connection first `getDefaultNamespace`
* Updated based on review
* Added tests

lstocchi reviewed (commented):
I didn't see the original commit before it was updated so if what i'm saying was the original behavior and it has been changed for some reason, forget about it.

By reading Jeff's comment it makes me think he was referring to having the kubernetes check connection function called inside the kube client class. So you just call listPods, it checks the connection and returns the right value (empty array or a reject or whatever if there is an error). So the UI just call listPods and that's it. Without having to know that it first have to check the connection and then execute listPods.

BTW i tested it and it does its job right 👍

cdrage commented:
> I didn't see the original commit before it was updated so if what i'm saying was the original behavior and it has been changed for some reason, forget about it.
> 
> By reading Jeff's comment it makes me think he was referring to having the kubernetes check connection function called inside the kube client class. So you just call listPods, it checks the connection and returns the right value (empty array or a reject or whatever if there is an error). So the UI just call listPods and that's it. Without having to know that it first have to check the connection and then execute listPods.
> 
> BTW i tested it and it does its job right 👍

Thanks! Yup, that's all intentional.

If we "check connection" and return an empty array for Kuernetes, we will never get an error saying that we're connection refused / unable to connect.

So I did not yet put any code more internally for connection rejection errors.

What do you think of the new changes @jeffmaury ?

benoitf commented on packages/main/src/plugin/kubernetes-client.ts:
```diff
@@ -271,7 +271,8 @@ export class KubernetesClient {
     const currentContext = this.kubeConfig.contexts.find(context => context.name === this.currentContextName);
-    if (currentContext) {
+    // Only update the namespace if we're able to actually connect to the cluster, otherwise we'll end up with a connection error.
+    if (currentContext && (await this.checkConnection())) {
```
I think it'll be more readable if the await is in a separate variable

benoitf commented on packages/main/src/plugin/kubernetes-client.ts:
```diff
@@ -423,6 +424,19 @@ export class KubernetesClient {
 
+  // Check that we can connect to the cluster and return a Promise<boolean> of true or false depending on the result.
+  // We will check via trying to retrieve a list of API Versions from the server.
+  async checkConnection(): Promise<boolean> {
```
maybe return a more complex object
because client might be interested in having the code

cdrage replied:
I've opened up this issue here: https://github.com/containers/podman-desktop/issues/2204 to keep this PR simple and short, we can implement this in a different PR.

benoitf commented on packages/renderer/src/stores/pods.ts:
```diff
@@ -22,11 +22,16 @@ import type { PodInfo } from '../../../main/src/plugin/api/pod-info';
-    podsInfos.set(result);
+  podsInfos.set(result);
+
+  if (await window.kubernetesCheckConnection()) {
```
I would avoid await in a if section

benoitf reviewed (changes requested):
I would expect to have kubernetesListPods always working (returning an empty array if I'm not able to connect)

If we ever need that errors in the UI

then `kubernetesListPods` should return a more complex objects like
```json
{
  "errors" : ["list", "of", "errors"],
  "pods": [pod1, pod2]
}
```
but for now I think client is just interested in having pods.
other commands could be interested in checking the status of the kubernetes cluster

for example in the dropdown of kube context, we could add an item saying that the connection is not working, etc but from store perspective we should not have the if clause

cdrage commented:
> I would expect to have kubernetesListPods always working (returning an empty array if I'm not able to connect)
> 
> If we ever need that errors in the UI
> 
> then `kubernetesListPods` should return a more complex objects like
> 
> ```json
> {
>   "errors" : ["list", "of", "errors"],
>   "pods": [pod1, pod2]
> }
> ```
> 
> but for now I think client is just interested in having pods. other commands could be interested in checking the status of the kubernetes cluster
> 
> for example in the dropdown of kube context, we could add an item saying that the connection is not working, etc but from store perspective we should not have the if clause

Updated PR with changes, as mentioned, I've opened up another issue that details that we should outline any connection  issues / inform the user: https://github.com/containers/podman-desktop/issues/2204

benoitf commented:
@cdrage I'm not sure to get the new changes.
The store is still pinging the kubernetes client to see if there are errors while it seems it's the kubernetes client that should return empty array in that case

cdrage commented:
> @cdrage I'm not sure to get the new changes. The store is still pinging the kubernetes client to see if there are errors while it seems it's the kubernetes client that should return empty array in that case

Updated and moved the logic to listPods within the Kubernetes client instead

jeffmaury reviewed (commented):
This add an additional round trip to the server when we get the pods why don't we simply fetch the pods and catch the error if it fails ?

cdrage commented:
> This add an additional round trip to the server when we get the pods why don't we simply fetch the pods and catch the error if it fails ?

You're right, it's an additional round trip. Although the call is so fast, it doesn't affect it much.

Having the `checkConnection` though would help us keep the code clean so instead we're doing `checkConnection()` rather than doing:

```
try {
 const pods = await this.listNamespacedPod(ns);
} catch(error) {
 if (error.message.includes('ECONNREFUSED') {
  // Ignore
  } else {
  throw error
  }
}
```

throughout all the code

lstocchi reviewed (approved):
LGTM

I think that as a further enhancement (another PR) we should show a notification if there is an error when calling the cluster. Otherwise if i'm sure there is a pod on my cluster and i see an empty list on PD i could think it's a PD issue.

cdrage commented:
Rebased and repushed 👍
